### ISSUE TITLE
MFhelper moving to act failures - improvement suggestion

### ISSUE THREAD

**Issue by DarkHorseDre** [kolton/d2bot-with-kolbot#1069]
I was using mfhelper to test something instead of tristramleecher and this happened:

leader goes from act 1 to act 4 for pots, followers are level 4 and havent been rushed yet.  they fail to take wp as below, then stall and exit game.

There is no need to exit as the player was coming back to act 1 to start the run (if it was never coming back, toons would wait until end, but thats as per config.  quitlist would also ensure they left)

Solution:  change code to not exit if act cannot be accessed, just continue wait loop for next command.

![image](https://user-images.githubusercontent.com/41862641/51409548-08ca6800-1b5a-11e9-8eda-c28375652a39.png)

**Comment by aim2kill** [comment]
Probably you may need to add a delay or commands to send the helper to the act you need him to be, but im pretty sure mfhelper is wo4king as intented 

MainLoop:
	while (true) {
		if (player) {
			while (!player.area) {
				delay(100);
			}

			playerAct = this.getPlayerAct(player);

			if (playerAct && playerAct !== me.act) {
				Town.goToTown(playerAct);
				Town.move("portalspot");
			}
Maybe a delay 10-12 sec to changes act

**Comment by DarkHorseDre** [comment]
no i think its more normal difficulty where act 1 pots are minor, and act 4 are the grand/super.
check town.js - it goes to highest act possible.  this is an issue at earlier levels as pots in act 4 are expensive, so I changed mine.

the bots currently track leader and goes to acts.  if the act they goto is unavailable, they quit.

what needs to change here is the bots don't quit!  some extra stuff could be added to work out if the act is needed for the run or not, but ultimately, quitting before working this out is premature as the leader inevitably is going to buy pots and return to lower act.

this is relevant in rushes, mfleader, etc.  this is not an issue in nm and hell as you mention, due to pot availability and costs being affordable by then.

**Comment by aim2kill** [comment]
I guess if you remove the towns from the getPlayerAct function it should do the trick, something like that:

	this.getPlayerAct = function (player) {
		if (player.area > 1 && player.area <= 39) {
			return 1;
		}

		if (player.area > 40 && player.area <= 74) {
			return 2;
		}

		if (player.area > 75 && player.area <= 102) {
			return 3;
		}

		if (player.area > 103 && player.area <= 108) {
			return 4;
		}

		if (player.area > 109) {
			return 5;
		}

		return false;
	};

**Comment by DarkHorseDre** [comment]
ah I see what you did not @5noop  - ignore the camp area id's :)

:+1:

**Comment by 5noop** [comment]
@aim2kill I already removed the towns ^^

**Comment by aim2kill** [comment]
Explain cause area 1,40,75,103 and 109 are town

**Comment by aim2kill** [comment]
So follower leave the game? after the change?

**Comment by DarkHorseDre** [comment]
i havent tried it but it will do - his fix avoids it trying to change acts if the bot goes to another TOWN in anothe act, but if it were to run a script in another act it would still cause them to exit as they would travel to the town of that act and wait at tp spot.

**Comment by aim2kill** [comment]
Yeah the only way I see is to add a command for the leader to sent to the mfhelper  to switch act when is really need it

**Comment by ryancrunchi** [comment]
I think the problem of trying to take a wp that is not accessible is more general.
The script breaks because, technically, attempting to take an inaccessible wp throws an error. Why ? This is not an error itself, it is just unable to do so. The function to take a wp would not throw an error, but only return false if the wp is inaccessible. This would make the current script continue and not break.
What do you think ?
Would changing a very used function break other cases ?

**Comment by DarkHorseDre** [comment]
yes thats exactly it!

I think the error is probably triggered outside of mfhelper (I didn't look but assume so) but as you say, it shouldn't fail and cause a complete quit.

now that's why I raise this as an enhancement request as validating across all scripts is required!

I had a quick think and I cannot think of a script that is required to exit game if leader is in act they cannot access, but then I have not used every script!  e.g. muling etc..

**Comment by DarkHorseDre** [comment]
@aim2kill that would actually be a bullet proof way of doing it if changing it globally broke other use cases!

**Comment by ryancrunchi** [comment]
The error is thrown is `Town.goToTown` function where the condition is

```
if (act !== me.act) {
	try {
		Pather.useWaypoint(towns[act - 1], wpmenu);
	} catch (WPError) {
		throw new Error("Town.goToTown: Failed use WP");
	}
}
```

In order to avoid throwing an error, the code would be replaced by
```
if (act !== me.act) {
	if (!Pather.useWaypoint(towns[act - 1], wpmenu)) {
		return false;
	}
}
```

And this will change the behavior for all calls to this function.

**Comment by ryancrunchi** [comment]
Tagging @noah- because this behavior changes need more discussion and it is currently lost in that 600-ish issues list...

**Comment by aim2kill** [comment]
Another Idea is to check for quest flag
A1 to A2 - andy is dead
A2 to A3 - duriel is dead
A3 to A4 - mepth is dead
A4 to A5 diablo is dead
If cant change stay in our current act

**Comment by DarkHorseDre** [comment]
@aim2kill you dont need both, the current method of checking access to wps (Acts) is sufficient (quest is irrelevant if next act town wp is inaccessible).

@ryancrunchi exactly - the error causes the issue (I assume throwerror forces exit, or?) - so an elegant exit in the **MFhelper use case** is whats needed - I just don't know if there's a use case where you want to quit?  like torchsystem?  even then, you're not following a leader so...

**Comment by aim2kill** [comment]
My bad , if quest isnt done then stay in the current act is what I mean to say

**Comment by ryancrunchi** [comment]
It is a question of semantic : not having access to a wp is not an error and should not stop the script (@DarkHorseDre throwing an error actually stops the script). It is just a use case that should be handled at scripts level, not common lib, IMHO.

**Comment by DarkHorseDre** [comment]
@ryancrunchi you are effectively agreeing with me no?  if wp checking is similar but accurate why write new code to do something else?  

its better to check absolutes (wps) than variables like quests where npc interaction and levels also need to be considered. especially if a player completes a quest but doesn't speak to jeryn/meshif/warriv/tyreal/walk through red portal (not sure if game autocompletes those?  I know quests can fail in some cases).

of course throwing an error is a problem, that's why I reported it :)

**Comment by ryancrunchi** [comment]
https://github.com/kolton/d2bot-with-kolbot/pull/1144
